src/pip/_internal/req/constructors.py

```diff
@@ -200,6 +200,59 @@ def install_req_from_editable(
     )
 
 
+def _looks_like_path(name):
+    # type: (str) -> bool
+    """Checks whether the string "looks like" a path on the filesystem.
+
+    This does not check whether the target actually exists, only judge from the
+    apperance. Returns true if any of the following is true:
+
+    * A path separator is found (either os.path.sep or os.path.altsep).
+    * The string starts with "." (current directory).
+    """
+    if os.path.sep in name:
+        return True
+    if os.path.altsep is not None and os.path.altsep in name:
+        return True
+    if name.startswith('.'):
+        return True
+    return False
+
+
+def _get_path_to_url(path, name):
+    # type: (str, str) -> str
+    """
+    First, it checks whether a provided path is an installable directory
+    (e.g. it has a setup.py). If it is, returns the path.
+
+    If false, check if the path is an archive file (such as a .whl).
+    The function checks if the path is a file. If false, if the path has
+    an @, it will treat it as a PEP 440 URL requirement and return the path.
+    """
+    if _looks_like_path(name) and os.path.isdir(path):
+        if is_installable_dir(path):
+            return path_to_url(path)
+        raise InstallationError(
+            "Directory %r is not installable. Neither 'setup.py' "
+            "nor 'pyproject.toml' found." % name
+        )
+    if not is_archive_file(path):
+        return None
+    if os.path.isfile(path):
+        return path_to_url(path)
+    urlreq_parts = name.split('@', 1)
+    if len(urlreq_parts) >= 2 and not _looks_like_path(urlreq_parts[0]):
+        return None
+    logger.warning(
+        'Requirement %r looks like a filename, but the '
+        'file does not exist',
+        name
+    )
+    return path_to_url(path)
+    # If the path contains '@' and the part before it does not look
+    # like a path, try to treat it as a PEP 440 URL req instead.
```

Comment:
It looks like you want to move this comment before the `return None` you just moved, since it goes along with that case (after the `if len(urlreq_parts) >= 2 and not _looks_like_path(...`).

Reply:
@cjerdonek I believe I forgot about this one. Sorry, I'll change it right now.